src/background_thread.c

```diff
@@ -350,24 +350,37 @@ check_background_thread_creation(tsd_t *tsd, unsigned *n_created,
 		}
 		background_thread_info_t *info = &background_thread_info[i];
 		malloc_mutex_lock(tsd_tsdn(tsd), &info->mtx);
-		if (info->started) {
-			pre_reentrancy(tsd);
-			int err = pthread_create_wrapper(&info->thread, NULL,
-			    background_thread_entry, (void *)(uint64_t)i);
-			post_reentrancy(tsd);
-
-			if (err == 0) {
-				++*n_created;
-				created_threads[i] = true;
-			} else {
-				malloc_printf("<jemalloc>: background thread"
-				    "creation failed (%d).\n", err);
-				if (opt_abort) {
-					abort();
-				}
+		bool create = info->started;
+		malloc_mutex_unlock(tsd_tsdn(tsd), &info->mtx);
+		if (!create) {
+			continue;
+		}
+
+		/*
+		 * To avoid deadlock with prefork handlers (which waits for the
+		 * mutex held here), unlock before calling pthread_create().
+		 */
+		malloc_mutex_unlock(tsd_tsdn(tsd),
+		    &background_thread_info[0].mtx);
+		pre_reentrancy(tsd);
+		int err = pthread_create_wrapper(&info->thread, NULL,
+		    background_thread_entry, (void *)(uint64_t)i);
+		post_reentrancy(tsd);
+		malloc_mutex_lock(tsd_tsdn(tsd),
+		    &background_thread_info[0].mtx);
+
+		if (err == 0) {
+			++*n_created;
+			created_threads[i] = true;
+		} else {
+			malloc_printf("<jemalloc>: background thread"
+				      "creation failed (%d).\n", err);
```

Comment:
Formatting and typo:

```C
malloc_printf("<jemalloc>: background thread "
    "creation failed (%d).\n", err);
```

BTW, we don't typically put periods at the end of error message strings.